# Python Basics: Coding Challenge
---
**This assignment is designed to provide you practice with variables, calculations, input, output, and data types.** 

## INSTRUCTIONS - Please read carefully.

1. Please complete the following coding exercises and upload a copy of the completed notebook as your assignment submission.

2. The estimated difficulty level is listed next to each question (**L** - low, **M** - medium, **H** - high).
   
3. A blank code cell has been created below each question for you to add the relevant code. Add additional code cells if needed. Execute all code cells to display the results. 

4. Add comments in each cell to explain what the code is meant to achieve (you will be graded on this). For example,

```
# ask for user preference

# check if user wants to continue
```

5. Follow guidelines discussed in class to create well-written code. For example, provide user friendly prompts for all user inputs and user friendly messages with all outputs. Assign meaningful names to variables (you will be graded on this).

6. <font color="red">The **scope of this assignment** is limited to the material covered until the end of the lesson on **<u>Python Fundamentals - 3: Debugging Code</u>**. You must not use concepts discussed in subsequent lessons in your work for this assignment</font>.

7. You can use additional materials, online resources and Gen AI tools to study course concepts but you may not post the specific questions online or otherwise seek help in answering the specific homework questions. Additionally, you must only use the concepts/ functions that are in the scope of the assignment and have been covered in the course. **The use of concepts and functions that are not in the scope of the assignment, or have not been covered in the course, or mentioned in the question will incur a <font color="red">penalty</font>**. The default penalty is zero for the question where such code is used, unless otherwise specified in the assignment. If we notice the use of concepts and functions that were not covered in the course or mentioned in the question, or if the code does not appear similar to what you have turned in before, we will ask you to set up a meeting to explain the code. **Failure to explain your submitted code may result in <font color="red">academic integrity violation proceedings</font>**. Please **CHECK THE ACADEMIC INTEGRITY SECTION IN THE SYLLABUS** for complete information. 
   
8. Do the following before downloading the notebook for submission. From the Kernel menu, select Restart Kernel and Clear All Outputs. Next, run all your code cells one by one, from the top to the bottom. Provide inputs where you have used input() functions and check that the outputs of all code cells are as you expected. If everything checks out, save the notebook and download it for submission. If you find any errors, fix them and repeat this step until everything checks out.

<HR/>

## <font color='red'/>Student Name</font>

#### **Name: Carly Caiazzo**

<HR/>

## Notebook Setup (Imports)

Please add any imports you need below or any other code that you need to run before your other code cells.


In [2]:

import math

# Take the square root of 4
print(math.sqrt(4))

#rounds to the smallest integer greater than or equal to 10.5
print(math.ceil(10.5))

#exponential function
print(math.pow(2,4))

2.0
11
16.0


## Question 1 (M)

Write a program that accepts the unit price and cost for a cup of lemonade at a lemonade stand as inputs and displays the profit in dollars and cents (X dollars and Y cents). Use the scenarios (called test cases) listed below to test your code. Run your code multiple times. Each time, set the inputs using the unit price and unit cost provided in the next test case and verify that the output your code gives matches the output listed in that test case. 

| Unit Price | Unit Cost | Profit in dollars and cents (Output) |
---------|--------|--------|
| 2.50 | 1.25 | 1 dollars and 25 cents |
| 3.23 | 1.23 | 2 dollars and 0 cents |
| 4.10 | 2.60 | 1 dollars and 50 cents (1 dollars and 49 cents is also correct) |
| 0.60 | 1.50 | 0 dollars and -90 cents |

Note: You may end up with 49 cents instead of 50 cents for the third test case above. That is fine. It is due to the way computers perform floating point arithmetic. While they are outside the scope of our course, Python programmers use `decimal` and `fractions` modules to deal with imprecision in floating point arithmetic.

In [42]:
# asks user for possible prices and costs per unit
unit_price = float(input("What is the unit price for a cup of lemonade?"))
unit_cost = float(input("What is the unit cost for a cup of lemonade?"))

#defines profit by subtracting cost from price
profit = (unit_price - unit_cost)

#defines dollar to cent conversion
cent_conversion = 100

#defines profit as cents
total_cents = round(profit * cent_conversion)

#separates dollars from cents
profit_dollars = int(total_cents / cent_conversion)
profit_cents = total_cents - (profit_dollars * cent_conversion)

#displays output in long-written format
print(profit_dollars, "dollars and", profit_cents, "cents")

What is the unit price for a cup of lemonade? 0.60
What is the unit cost for a cup of lemonade? 1.50


0 dollars and -90 cents


<HR/>

## Question 2 (L)

Below is a program that prompts a user for two numbers as inputs and stores them in variables num1 and num2. The program will then swap the values in variables num1 and num2. Finally the program outputs the swapped values.

Below is some code pre-filled to accomplish this task with some blank spots. Fill in the blank spots and then run the code to confirm it works correctly.

**Note:** To get full credit on this question, the values stored in num1 and num2 must be swapped.

In [31]:
### Prompt the User for the numbers to store in num1 and num2
num1 = int(input("What is the first number to store in num1?"))
num2 = int(input("What is the second number to store in num2?"))


# Use a temporary variable (temp) to hold one of the numbers in order to swap
print("Swapping numbers...")
num2 = num1
num1 = temp
temp = num1

# Output the new stored numbers
print("The number stored in num1 is now", num1)
print("The number stored in num2 is now", num2)

What is the first number to store in num1? 1
What is the second number to store in num2? 2


Swapping numbers...
The number stored in num1 is now 2
The number stored in num2 is now 1


<HR/>

## Question 3 (H)

A retailer has installed a new section for kids' toys. There are five shelving units in this section, each with four shelves. Each shelf is ten feet long. Three of the shelving units will be used to store small toys and two shelving units will be used to store large toys. Each small toy takes up eleven inches space in length (package + the space needed between two toys). Each large toy takes up fifteen inches.

Write a program that captures the above information and calculates how many small toys and how many large toys can be stored in the new section. 

**Note:** Make sure to think about which variables should be CONSTANTS and which should be regular variables when writing code.

**Hint:** If your calculations are correct, your code should produce the following answer:

```
120 toys can fit on the shelving units for SMALL toys.
64 toys can fit on the shelving units for LARGE toys.
```

In [28]:
# feet per shelf, according to which size toy
shelf_smalltoys_feet = 4*10
shelf_largetoys_feet = 4*10

#converting shelf feet to inches
shelf_smalltoys_inches = shelf_smalltoys_feet * 12
shelf_largetoys_inches = shelf_largetoys_feet * 12

#toy size, with .5 inches of shelf space on each side for small toys
smalltoys = 11 + .5*2
largetoys = 15

#amount of toys that can be placed on each shelf, according to size
small_total = (shelf_smalltoys_inches/smalltoys)*3
large_total = (shelf_largetoys_inches/largetoys)*2

#shows output
print (int(small_total), "toys can fit on the shelving units for SMALL toys")
print (int(large_total), "toys can fit on the shelving units for LARGE toys")

120 toys can fit on the shelving units for SMALL toys
64 toys can fit on the shelving units for LARGE toys


<HR/>

## Question 4 (M)

Write a program that prompts the user for the number of seconds. The program will then calculate and print the number of days, hours, minutes, and seconds the entry is equivalent to.

**Note:** You can assume the number of seconds being entered will always be an integer value.

Test your code using the following test cases:

| seconds | Output (days, hours, minutes, and seconds) |
---------|--------|
| 60 | 0 days, 0 hours, 1 minutes, and 0 seconds |
| 3600 | 0 days, 1 hours, 0 minutes, and 0 seconds | 
| 23455  | 0 days, 6 hours, 30 minutes, and 55 seconds |
| 86400  | 1 days, 0 hours, 0 minutes, and 0 seconds |
| 846600 | 9 days, 19 hours, 10 minutes, and 0 seconds |


In [1]:
# asks user for a hypothetical number of seconds
seconds_input = int(input("Give me a possible number of seconds"))

#defines days
days = seconds_input//(24*60*60)
remaining_seconds = seconds_input%(24*60*60)

#defines hours
hours = remaining_seconds//(60*60)
remaining_seconds = remaining_seconds%(60*60)

#defines minutes
minutes = remaining_seconds//(60)

#defines seconds
seconds = remaining_seconds%60

#displays output
print(days, "days," , hours, "hours," , minutes, "minutes, and" , seconds,"seconds")

Give me a possible number of seconds 846600


9 days, 19 hours, 10 minutes, and 0 seconds


<HR/>

## Mini-Lesson on python `imports`

Sometimes you will need to use some more advanced mathematical operations besides beyond `+, -, /, *, and %`. In those cases, python has a built in library for providing extra features such as power, square root, sin, cos, tan, etc. To read about all the different features of the `math` library you can visit the [python math library documentation page](https://docs.python.org/3/library/math.html) and read through the different functions and how to use them. For this homework assignment it would be good to familiarize yourself with the `sqrt`, `ceil`, and `pow` functions.

In order to use any of the functions from the `math` library, you will need to import the `math` library. You can do this once at the beginning of your notebook in the `Notebook Setup (Imports)` section (just make sure you run that cell first) or you can put it at the start of every cell where you use math (there's no harm in importing the same library multiple times).

See the code cell below for an example of how to import and use the math library for the `sqrt` function. When you run it, you should get 2 since that is the sqrt of 4. Try changing the number to 16, 25, 36 and see what the results are.


In [23]:
import math

# Take the square root of 4
print(math.sqrt(4))

print(math.pow(2,2))

2.0
4.0


<HR/> 

## Question 5 (M)

Please write a program that uses the [quadradic formula](https://en.wikipedia.org/wiki/Quadratic_formula) to calculate solution(s) to the quadratic equation. 

$$x = \frac{-b \pm \sqrt{b^2 - 4ac}}{2a}$$

Your program should:
- prompt the user for the values a, b, and c
- calculate the two solutions
- print the two solutions back to the user rounded to two decimal places

**Note:** For simplicity, you can assume that the coefficients (a,b,c) will only be integers and will always yield real numbers solutions.

Test your code with the following test cases:

| a | b | c | Solutions (output) |
---------|--------|--------|--------|
| 1 | 3 | -40 | 5, -8 |
| 2 | 2 | -420 | 14, -15 |
| 1 | 7 | 4 | -0.63, -6.37 |
| -8 | 12 | 6 | -0.40, 1.90 |

In [45]:
# asks user for input values a, b, c
a = int(input("Give me a possible value for a"))
b = int(input("Give me a possible value for b"))
c = int(input("Give me a possible value for c"))

#defines discriminant and states that it can be used if positive
discriminant = math.pow(b,2) - (4 * a * c)
useable_discriminant = abs(discriminant)

#completes formula in two parts for addition and subtraction
add_solution = (-b + math.sqrt(useable_discriminant))/(2*a)
subtract_solution = (-b - math.sqrt(useable_discriminant))/(2*a)

#displays both outputs as rounded decimals
print(round(add_solution, 2), ",", round(subtract_solution, 2))

Give me a possible value for a 1
Give me a possible value for b 7
Give me a possible value for c 4


-0.63 , -6.37


<HR/>

## Question 6 (H)

You're working on a video game and you are given a damage calculation formula shown below to implement in code.

$$Damage = \frac{(\frac{2 * Level * Power}{5} + 2) * \frac{Attack}{Defense}}{50} + 2^{Critical}$$

Create variables for the following values in the formula and assign them values:
- Attack
- Defense
- Level
- Critical
- Power

Then implement the formula in code to calculate the damage and output to the user how much damage was done. Round your answer to 2 decimal places.

Test your code using the following values:


| Attack | Defense | Level | Critical | Power | Damage (Output) |
---------|--------|--------|--------|--------|--------|
| 25 | 50 | 30 | 1 | 50 | 8.02 |
| 40 | 25 | 80 | 1 | 20 | 22.54 |
| 10 | 20 | 8 | 2.5 | 5 | 5.84 |
| 41 | 36 | 54 | 1 | 25 | 14.35 |


In [20]:
# creating and defining variables
attack = 40
defense = 25
level = 80
critical = 1
power = 20

#implimenting formulas into code
damage = (((((2*level*power)/5 )+2) * (attack/defense))/50) + math.pow(2,critical)

#displays output values
print(round(damage,2))

22.54


<HR/>

## Question 7 (L)

Your friend comes to you with a problem with their code. The code is supposed to calculate the average of a number of variables. They said it used to work and they made some changes, but now it no longer works. Help them fix the code after copying it to the code cell below and use comments to explain what was the issue.

```Python
NUMBER_1 = 10
NUMBER_2 = 15
NUMBER_3 = 20
NUMBER_4 = 25
NUMBER_5 = 30
NUMBEr_6 = 35

SUM = NUMBER_1 + NUMBER_2 + NUMBER_3 + NUMBER_4 + NUMBER_5 NUMBER_6

AVG = TOTAL / 5

print("The average is " + str(avg))
```

The output should be `The average is 22.5`.

In [8]:
# defining number variables
#ISSUE FIXED - "NUMBEr_6" should be "NUMBER_6" for later use to find sum of variables
NUMBER_1 = 10
NUMBER_2 = 15
NUMBER_3 = 20
NUMBER_4 = 25
NUMBER_5 = 30
NUMBER_6 = 35

#adding all variables
#ISSUE FIXED - does not have an addition function (+) between NUMBER_5 and NUMBER_6
SUM = NUMBER_1 + NUMBER_2 + NUMBER_3 + NUMBER_4 + NUMBER_5 + NUMBER_6

#finding mean or average 
# ISSUE FIXED - "TOTAL" not defined variable, changed to "SUM" which has been defined
# ISSUE FIXED - should be divided by 6 to account for 6 variables
AVG = SUM / 6

#displaying output
#ISSUE FIXED - "avg" is not capitalized, only "AVG" has been defined
print("The average is " + str(AVG))

The average is 22.5


<HR/> 

## Question 8 (M)

**Trace Tables should be completed as tables in Markdown. Please review the lessons on `Thinking like a Programmer` for instructions on creating a table in Markdown.**
**If you run into issues and cannot resolve them in time, you may create the trace table on paper and take a picture or in Excel and take a screenshot. When submitting in CodeGrade, include the image as an additional file with along with your notebook submission. Make sure to annotate below that you submitted it as an image.**


Create a trace table for the following code. You can create a table using markdown in the markdown cell provided below or you can create the table on paper or another tool and attach an image in the markdown cell. See posted class material and resources for help with creating tables and attaching images.



```Python
a = 25

b = 25

a / 5

a = a / 5

c = b + a * 2

(c + c) / 10

b = a + c // b

a = c % b

b = c

a = b
```



| Line | a  | b  | c  |
|------|----|----|----|
| 1    | 25 |    |    |
| 2    |    | 25 |    |
| 4    | 5  |    |    |
| 5    |    |    | 35 |
| 7    |    | 1  |    |
| 8    | 0  |    |    |
| 9    |    | 35 |    |
| 10   | 35 |    |    |

<HR/>

## Question 9 (L)

Create a trace table for the following code. Feel free to look up online what each of the functions used below does.

```Python
import math

a = math.ceil(2.5)

b = a * 3

c = (a + b) // 5

b = max(a - c, c - a)

a = math.sqrt(a + b)

b = math.pow(c, 2)

print(a)

print(b)

print(c)


```


| Line | a | b | c | Output |
|------|---|---|---|--------|
| 2    | 3 |   |   |        |
| 3    |   | 9 |   |        |
| 4    |   |   | 2 |        |
| 5    |   | 1 |   |        |
| 6    | 2 |   |   |        |
| 7    |   | 4 |   |        |
| 8    |   |   |   | 2      |
| 9    |   |   |   | 4      |
| 10   |   |   |   | 2      |

<HR/>

## Question 10 (H)

Your friend is having issues with their jupyter notebook. They have 3 code cells set up, each with a specific task. They are running into errors and they're not sure what happened. Help them fix the errors in their notebook and try to figure out what the problem was.

Some of the problems they're experiencing:
- When they re-opened the file again, the first code block no longer runs. Instead, the code block gives an error.
- The second code block runs OK, but it does not give the correct answer.
- Each time the third code block runs, it gives a different answer.

In [10]:
#define SIDE varibles

SIDE_1 = 3
SIDE_2 = 4
SIDE_3 = 5

## Calculate the perimeter of a triangle
perimeter = SIDE_1 + SIDE_2 + SIDE_3

print("The perimeter is: " + str(round(perimeter)))

The perimeter is: 12


In [12]:
# Triangle sides
SIDE_1 = 3
SIDE_2 = 4
SIDE_3 = 5

## Calculate the area of a square

# Square Side
SIDE = 10

# Square Area
area = SIDE*SIDE

print(f"The area of the square with side {SIDE} is: {area}")

# Cylinder Dimensions height 10, base diameter 16
HEIGHT = 10 
BASE   = 16

The area of the square with side 10 is: 100


In [17]:
# Calculate volume of a cylinder 
import math

## Calculate the area of the base (pi * r^2)

#define BASE variable
BASE = 16

#calculate area of base
AREA = (BASE/2)**2 * math.pi

volume = HEIGHT * BASE

print("The area is : " + str(round(AREA,2)))

The area is : 201.06


Summary: 

Code Cell 1:
    - ISSUE: all 3 SIDE variables were never defined
    - FIX: defined SIDE_1, SIDE_2, SIDE_3

Code Cell 2: 
    -ISSUE: using varibles defined for the sides of a triangle to calculate the area of a square
    -FIX: defined area variable as SIDE*SIDE, since those were the defined variables for squares

Code Cell 3: 
    -ISSUE: BASE variable never defined 
    -ISSUE: need two separate variables (can't have BASE = BASE/2...)
    -FIX: define BASE and then calculate area by creating new variable AREA that follows the area of a cylinder equation
    -ISSUE: specific task says to "calculate the area of base (pi * r^2), but the printed output shows volume
    -FIX: print(AREA) istead

<HR/>

### <font color = 'blue'>SCOPE CHECK</font> 
##### Check the work you are turning in to ensure that
- You haven't used concepts or functions that are covered in lessons that are outside the scope of this assignment (scope provided in instructions above).
##### <font color = 'blue'> Using concepts or functions that are outside the scope of the assignment will result in a score of zero for the corresponding questions.</font>


### <font color = 'red'>ACADEMIC INTEGRITY CHECK</font> 

##### Check the work you are turning in to ensure that
- You haven't used any concepts or functions not explicitly covered in the course lessons. You cannot use concepts/functions available in Python documentation unless they have been explicitly covered in a lesson or the question prompt explicitly instructs you to do so.
- You haven't used concepts or functions that are covered in lessons that are outside the scope of this assignment (scope provided in instructions above).
- You haven't copied any code from any websites or AI assistant tools.
- You haven't used code provided to you by another individual (other than your teammates for team assignments).
- You can explain every line of code you are turning in.

##### <font color = 'red'> Failing to do the above could result in a formal academic integrity violation filing. </font>

If any potential AI violations are identified in your submission, it will not receive a grade. You will be asked to schedule a meeting with the instructor to explain the code that was flagged for a potential violation. If you are able to satisfactorily explain your work and any concepts, functions, etc. that were used, we will assign a grade for the assignment which will include a penalty for out-of-scope work. If you are unable to satisfactorily explain your work and any concepts, functions, etc. that were used, an AI violation form will be filed.

<HR/>
